### ISSUE TITLE
There is no good process for updating dashboards included in App plugins

### ISSUE THREAD

**Issue by MatMeredith** [grafana/grafana#6522]
I have an app plugin that defines a set of dashboards as JSON files.   This works well -- the user enables the plugin and all the dashboards get imported.   It's a single click operation for the user, if they try to make changes to the dashboard then they get warned that they are changing dashboards defined in the plugin etc.

The issue is what happens when I (as the developer) want to make changes to the dashboards.
- I make the changes working in the Web UI.
- Now I need to get updated JSON files to include in my plugin.

I had been doing this by using the HTTP API to get the dashboard JSON.  This doesn't work though now that I have repeated panels in my dashboards as exporting them in this way doesn't templatize the repeated panels.   

If I want repeated panels to be templatized then I have to export them through the Web GUI using the Share / Export option.   But the JSON generated in this way can't be used to define dashboards in the plugin.

There doesn't seem to be a solution that works for this key workflow?

[Grafana v3.1.1 (commit: v3.1.0+21-g3921cc7-dirty)]

**Comment by torkelo** [comment]
set a revision property in the dashboard json, 
example: 
https://github.com/grafana/grafana/blob/master/public/app/plugins/app/testdata/dashboards/graph_last_1h.json#L2

Also update the version of the app plugin. When someone then update's the app plugin and restart grafana server, it will detect a new app version has been installed and look for updated dashboard (based on installed revision), and automatically update the installed dashboards

**Comment by MatMeredith** [comment]
Sorry - I obviously wasn't clear, but I don't see how that addresses my issue?   I have no issue with versioning.   

Let me try to explain again.   
- I have a Grafana app plugin containing a set of dashboards.
- I want to make enhancements to the dashboards.   So I install my plugin on a Grafana server, import the dashboards and make a set of changes using the Web GUI.   Adding new panels etc.
- I now want to distribute an updated version of my plugin with the enhancements I have made.  That means that I need to update the JSON files in my app plugin directory.
- But there is no way to generate that JSON from the updated dashboards on my Grafana server.   If I export them via the Web UI then they are templatized in a way that is not suitable for use in an app plugin.   If I export them via the HTTP API then repeated panels don't get templatized.

Does that make sense?

**Comment by torkelo** [comment]
not sure why View JSON would not work. 

>  If I export them via the HTTP API then repeated panels don't get templatized.

Not sure I understand this. You want them to be templatized but not in the way they are with the export logic?

**Comment by MatMeredith** [comment]
Because (unless I am mistaken) the JSON that is generated in that way is not compatible with the JSON that is needed for dashboard definitions in app plugins?   It fails to import when I enable the plugin?

**Comment by torkelo** [comment]
how does it fail?

- View JSON
- Copy JSON paste into app dashboard json file
- Remove id and set a revision property, 
- Update app version, 
- Restart grafana-server and it should update installed dashboad

**Comment by torkelo** [comment]
if you have template variables that contain sensitive data you might need to remove that as View JSON does not do any cleaning of the json model, like the export process do

**Comment by torkelo** [comment]
https://github.com/grafana/grafana/blob/master/public/app/core/time_series2.ts#L78

**Comment by MatMeredith** [comment]
Ah -- this is why we are talking at cross purposes I think.   When you said "View JSON" I thought you meant  Share Dashboard: Export: View JSON.   

Yes, the sensitive information is exactly why this process doesn't work.   Although in fact it isn't just "sensitive information" that is the issue.   The first time the dashboard is loaded on another system, if the query that populates the template variable does not yet return any data, the user sees all the instances of the repeated panel that existed at the time the dashboard was exported from my system.

The only option I see is to write custom scripts that strip out all recurrences of repeated panels and remove sensitive information from ScopedVars, but this doesn't feel like an issue that only I am going to need to solve...

**Comment by torkelo** [comment]
in this case you need to use the Export feature, then remove the templating it generates. We do not have  an automated process for exporting dashboards that can be used in apps (without any customizations)

**Comment by MatMeredith** [comment]
Can this issue be reopened then please as a feature request?   For anyone trying to distribute dashboards as app plugins this seems to me like a major limitation?
